src/v/ssx/single_fiber_executor.h

```diff
@@ -0,0 +1,324 @@
+        handle_t(const handle_t&) = delete;
+        handle_t& operator=(const handle_t&) = delete;
+        handle_t(handle_t&&) = delete;
+        handle_t& operator=(handle_t&&) = delete;
+        ~handle_t() = default;
+    };
+
+    struct running_t {
+        std::optional<handle_t> handle;
+
+        explicit running_t(promise_t&& promise)
+          : handle(std::move(promise)) {}
+        void run(Func&& func, ss::weak_ptr<self_t>&& executor) {
+            try {
+                ssx::background
+                  = func(*handle->as)
+                      .then_wrapped([executor = std::move(executor),
+                                     as = handle->as](future_t&& f) mutable {
+                          if (executor) {
+                              executor->complete_running(std::move(f));
+                          } else {
+                              // caller received errc::interrupted, no-one is
+                              // interested in the result
+                              f.ignore_ready_future();
+                          }
+                          return ss::now();
+                      });
+            } catch (...) {
+                vassert(executor, "executor is nullptr");
+                executor->complete_running(
+                  ss::make_exception_future<ret_value_t>(
+                    std::current_exception()));
+            }
+        }
+        running_t(const running_t&) = default;
+        running_t& operator=(const running_t&) = default;
+        running_t(running_t&&) = default;
+        running_t& operator=(running_t&&) = default;
+        ~running_t() { vassert(!handle, "promise not fulfilled"); };
+    };
+
+    struct requested_t {
+        Func func;
+        promise_t promise;
+
+        explicit requested_t(Func&& func)
+          : func(std::forward<Func>(func)) {}
+    };
+
+    // _running is set iff the background fiber is running
+    std::optional<running_t> _running;
+    std::optional<requested_t> _pending;
+
+    enum class status { operating, draining, drained };
+    status _status{status::operating};
+
+    void launch(promise_t&& promise, Func&& func) {
+        vassert(!_running, "launch() called while already running");
+        _running.emplace(std::move(promise));
+        _running->run(std::forward<Func>(func), weak_from_this());
+    }
+
+    void abort_running() {
+        if (_running && _running->handle) {
+            _running->handle->as->request_abort();
+            _running->handle->promise.set_value(
+              wrapped_ret_value_t{std::unexpect, errc::interrupted});
+            _running->handle = std::nullopt;
+        }
+        // keep _running set to indicate the fiber is still running
+    }
+
+    void complete_running(future_t&& f) {
+        vassert(_running, "no info about the background fiber completed");
+        if (_running->handle) {
+            auto& p = _running->handle->promise;
+            if (f.failed()) [[unlikely]] {
```

Comment:
future.hh has some forward_to stuff that might be less code and more efficient here

Reply:
these would be really nice to use, but unfortunately these future and promise have different types, see https://github.com/redpanda-data/redpanda/pull/27932/files/5feaf8de7a695d7e36d3df5dadace22c6907bee8#diff-6fc1806ad27d337fd2531aa6dfbdfbcdde53d7cf0bc35be29675cc2a733357a1R74